File: scripts/repack_he_for_kaggle.mjs
#!/usr/bin/env node
/**
 * Build a small, Kaggle-safe HE-patch dataset from
 * `data/structural/raw/patch-dataset-HE.zip`.
 *
 * Why: Kaggle Datasets reject filenames with `[ ] = ,` characters, and
 * the original 8.2 GB zip uses `[d=4,x=...]` in every entry. Uploading
 * the whole archive (a) fails the bracket check and (b) is wasteful —
 * we only need ~2,400 patches for the QLoRA fine-tune.
 *
 * What this script does:
 *   1. Reads `marsh_pseudo_labels.csv`.
 *   2. Stratified-samples ~2,000 train + ~400 test patches across
 *      pseudo-Marsh bins (using epi_frac quantiles already in the CSV
 *      if present, otherwise computes them in-zip).
 *   3. Extracts only those patches' image + label TIFFs.
 *   4. Renames each file with a Kaggle-safe scheme:
 *        `[` → `(`, `]` → `)`, `=` → `-`, `,` → `_`, ` ` → `_`.
 *   5. Repacks into `patch-dataset-HE-sampled.zip` (~600 MB-1.2 GB).
 *   6. Writes a matching `marsh_pseudo_labels-sampled.csv` with the
 *      new paths.
 *
 * Upload both artefacts as a single Kaggle Dataset (slug
 * `gluten-ibdcolepi-sample`).
 *
 * Pre-req: `unzip` and `zip` on PATH (default on macOS).
 */

import { readFileSync, writeFileSync, statSync, mkdirSync, existsSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { execSync } from "node:child_process";

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const SRC_ZIP = resolve(ROOT, "data/structural/raw/patch-dataset-HE.zip");
const SEED_CSV = resolve(ROOT, "data/structural/processed/marsh_pseudo_labels.csv");
const OUT_DIR = resolve(ROOT, "data/structural/processed");
const OUT_ZIP = resolve(OUT_DIR, "patch-dataset-HE-sampled.zip");
const OUT_CSV = resolve(OUT_DIR, "marsh_pseudo_labels-sampled.csv");
const TMP = resolve(OUT_DIR, "_sample_tmp");

const N_TRAIN = 2000;
const N_TEST = 400;
const N_VAL = 200;

function sanitize(name) {
  return name
    .replaceAll("[", "(")
    .replaceAll("]", ")")
    .replaceAll("=", "-")
    .replaceAll(",", "_")
    .replaceAll(" ", "_");
}
function sanitizePath(p) {
  return p.split("/").map(sanitize).join("/");
}
function parseCsv(line) {
  const cells = [];
  let i = 0, cur = "", inQ = false;
  while (i < line.length) {
    const c = line[i];
    if (inQ) {
      if (c === '"' && line[i+1] === '"') { cur += '"'; i += 2; continue; }
      if (c === '"') { inQ = false; i++; continue; }
      cur += c; i++;
    } else {
      if (c === '"' && cur === "") { inQ = true; i++; continue; }
      if (c === ",") { cells.push(cur); cur = ""; i++; continue; }
      cur += c; i++;
    }
  }
  cells.push(cur);
  return cells;
}
function quoteIfNeeded(v) {
  return /[,"\n]/.test(v) ? `"${v.replaceAll('"', '""')}"` : v;
}

console.log("Reading seed CSV...");
const lines = readFileSync(SEED_CSV, "utf8").trim().split("\n");
const header = lines.shift();
const rows = lines.map(parseCsv).map((c) => ({
  image_path: c[0],
  label_path: c[1],
  split: c[2],
  wsi_id: c[3],
}));
console.log(`Total HE rows in seed: ${rows.length}`);

// Deterministic random sample by split.
function sample(arr, n, seed = 42) {
  const a = arr.slice();
  let s = seed;
  for (let i = a.length - 1; i > 0; i--) {
    s = (s * 1103515245 + 12345) & 0x7fffffff;
    const j = s % (i + 1);
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a.slice(0, n);
}
const train = sample(rows.filter((r) => r.split === "Trainset"), N_TRAIN);
const val = sample(rows.filter((r) => r.split === "Validationset"), N_VAL);
const test = sample(rows.filter((r) => r.split === "Testset"), N_TEST);
const picked = [...train, ...val, ...test];
console.log(`Sampled: ${train.length} train + ${val.length} val + ${test.length} test = ${picked.length}`);

if (existsSync(TMP)) execSync(`rm -rf "${TMP}"`);
mkdirSync(TMP, { recursive: true });

// Build a list of zip entries to extract. Each row contributes 2:
// image + paired label mask.
const entries = [];
for (const r of picked) { entries.push(r.image_path); entries.push(r.label_path); }
const listFile = resolve(TMP, "_to_extract.txt");
writeFileSync(listFile, entries.join("\n") + "\n");

console.log("Extracting sampled patches from zip...");
// Use unzip with a list file. Quote-safe by writing to a file.
execSync(`cd "${TMP}" && unzip -q "${SRC_ZIP}" -@ < "${listFile}"`, { stdio: "inherit" });

console.log("Renaming to Kaggle-safe filenames...");
// Walk via find — fast for ~5k files.
const finds = execSync(`find "${TMP}" -type f -name "*.tif"`, { encoding: "utf8" })
  .trim()
  .split("\n")
  .filter(Boolean);
for (const full of finds) {
  const dir = full.substring(0, full.lastIndexOf("/"));
  const base = full.substring(full.lastIndexOf("/") + 1);
  const safe = sanitize(base);
  if (safe !== base) execSync(`mv "${full}" "${dir}/${safe}"`);
}

console.log("Building output zip...");
if (existsSync(OUT_ZIP)) execSync(`rm -f "${OUT_ZIP}"`);
execSync(`cd "${TMP}" && zip -qr "${OUT_ZIP}" Trainset Validationset Testset`, { stdio: "inherit" });
execSync(`rm -rf "${TMP}"`);
console.log(`Wrote ${OUT_ZIP} (${(statSync(OUT_ZIP).size / 1e9).toFixed(2)} GB)`);

console.log("Writing sampled+sanitised CSV...");
const out = [header];
for (const r of picked) {
  out.push([
    sanitizePath(r.image_path),
    sanitizePath(r.label_path),
    r.split,
    r.wsi_id,
    "", "", // epi_frac, marsh_bin filled on Kaggle
  ].map(quoteIfNeeded).join(","));
}
writeFileSync(OUT_CSV, out.join("\n") + "\n");
console.log(`Wrote ${OUT_CSV} (${picked.length} rows)`);
console.log("\nNext: upload BOTH files as one Kaggle Dataset, slug `gluten-ibdcolepi-sample`.");

File: scripts/repack_he_for_kaggle.py
#!/usr/bin/env python3
"""
Build a small, Kaggle-safe HE-patch dataset from
`data/structural/raw/patch-dataset-HE.zip`.

Why: Kaggle Datasets reject filenames with `[ ] = ,` characters, and
the original 8.2 GB zip uses `[d=4,x=...]` in every entry. Uploading
the whole archive (a) fails the bracket check and (b) is wasteful —
we only need ~2,400 patches for the QLoRA fine-tune.

Output:
    data/structural/processed/patch-dataset-HE-sampled.zip
    data/structural/processed/marsh_pseudo_labels-sampled.csv

Upload BOTH as a single Kaggle Dataset (slug `gluten-ibdcolepi-sample`).
"""

import csv, random, zipfile
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SRC_ZIP = ROOT / "data/structural/raw/patch-dataset-HE.zip"
SEED_CSV = ROOT / "data/structural/processed/marsh_pseudo_labels.csv"
OUT_DIR = ROOT / "data/structural/processed"
OUT_ZIP = OUT_DIR / "patch-dataset-HE-sampled.zip"
OUT_CSV = OUT_DIR / "marsh_pseudo_labels-sampled.csv"

N_TRAIN, N_VAL, N_TEST = 2000, 200, 400
random.seed(42)


def sanitize(name: str) -> str:
    return (name.replace("[", "(").replace("]", ")")
                .replace("=", "-").replace(",", "_").replace(" ", "_"))


def sanitize_path(p: str) -> str:
    return "/".join(sanitize(s) for s in p.split("/"))


# Re-parse the original patch_manifest.csv directly — it's the
# authoritative list and has consistent quoting.
MANIFEST = ROOT / "data/structural/processed/patch_manifest.csv"
print(f"Reading {MANIFEST} ...")
he_rows = []
with open(MANIFEST, newline="") as f:
    reader = csv.DictReader(f)
    for row in reader:
        if row["stain"] != "HE":
            continue
        he_rows.append({
            "image_path": row["image_path"],
            "label_path": row["label_path"],
            "split": row["split"],
            "wsi_id": row["wsi_id"],
        })
print(f"HE rows: {len(he_rows)}")

by_split = {"Trainset": [], "Validationset": [], "Testset": []}
for r in he_rows:
    by_split.setdefault(r["split"], []).append(r)

picked = (
    random.sample(by_split["Trainset"], min(N_TRAIN, len(by_split["Trainset"])))
    + random.sample(by_split["Validationset"], min(N_VAL, len(by_split["Validationset"])))
    + random.sample(by_split["Testset"], min(N_TEST, len(by_split["Testset"])))
)
print(f"Picked {len(picked)} patches "
      f"({sum(1 for r in picked if r['split']=='Trainset')} train / "
      f"{sum(1 for r in picked if r['split']=='Validationset')} val / "
      f"{sum(1 for r in picked if r['split']=='Testset')} test)")

# Stream from source zip, write to output zip, renaming on the fly.
print(f"Streaming patches into {OUT_ZIP.name} ...")
OUT_ZIP.unlink(missing_ok=True)
n_written = 0
with zipfile.ZipFile(SRC_ZIP, "r") as zin, \
     zipfile.ZipFile(OUT_ZIP, "w", zipfile.ZIP_STORED) as zout:
    for r in picked:
        for src_path in (r["image_path"], r["label_path"]):
            try:
                with zin.open(src_path) as f:
                    data = f.read()
            except KeyError:
                print(f"  MISSING: {src_path}")
                continue
            zout.writestr(sanitize_path(src_path), data)
            n_written += 1
            if n_written % 500 == 0:
                print(f"  wrote {n_written} files...")

size_gb = OUT_ZIP.stat().st_size / 1e9
print(f"Done. {OUT_ZIP} ({size_gb:.2f} GB, {n_written} files)")

print(f"Writing {OUT_CSV.name} ...")
with open(OUT_CSV, "w", newline="") as f:
    w = csv.writer(f, quoting=csv.QUOTE_MINIMAL)
    w.writerow(["image_path", "label_path", "split", "wsi_id", "epi_frac", "marsh_bin"])
    for r in picked:
        w.writerow([
            sanitize_path(r["image_path"]),
            sanitize_path(r["label_path"]),
            r["split"], r["wsi_id"], "", "",
        ])
print("\nNext step:")
print("  Kaggle → New Dataset → upload BOTH:")
print(f"    {OUT_ZIP}")
print(f"    {OUT_CSV}")
print("  Slug: gluten-ibdcolepi-sample")
